src/Microsoft.ML.StaticPipe/KMeansStatic.cs

```diff
@@ -29,20 +30,63 @@ public static class KMeansClusteringExtensions
         public static (Vector<float> score, Key<uint> predictedLabel) KMeans(this ClusteringContext.ClusteringTrainers ctx,
            Vector<float> features, Scalar<float> weights = null,
-           int clustersCount = KMeansPlusPlusTrainer.Defaults.K,
-           Action<KMeansPlusPlusTrainer.Arguments> advancedSettings = null,
+           int clustersCount = KMeansPlusPlusTrainer.Defaults.ClustersCount,
            Action<KMeansModelParameters> onFit = null)
         {
             Contracts.CheckValue(features, nameof(features));
             Contracts.CheckValueOrNull(weights);
             Contracts.CheckParam(clustersCount > 1, nameof(clustersCount), "If provided, must be greater than 1.");
             Contracts.CheckValueOrNull(onFit);
-            Contracts.CheckValueOrNull(advancedSettings);
 
             var rec = new TrainerEstimatorReconciler.Clustering(
             (env, featuresName, weightsName) =>
             {
-                var trainer = new KMeansPlusPlusTrainer(env, featuresName, clustersCount, weightsName, advancedSettings);
+                var options = new KMeansPlusPlusTrainer.Options
+                {
+                    FeatureColumn = featuresName,
+                    ClustersCount = clustersCount,
+                    WeightColumn = weightsName != null ? Optional<string>.Explicit(weightsName) : Optional<string>.Implicit(DefaultColumnNames.Weight)
+                };
+
+                var trainer = new KMeansPlusPlusTrainer(env, options);
+
+                if (onFit != null)
+                    return trainer.WithOnFitDelegate(trans => onFit(trans.Model));
+                else
+                    return trainer;
+            }, features, weights);
+
+            return rec.Output;
+        }
+
+        /// <summary>
+        /// KMeans <see cref="ClusteringContext"/> extension method.
+        /// </summary>
+        /// <param name="ctx">The regression context trainer object.</param>
+        /// <param name="features">The features, or independent variables.</param>
+        /// <param name="weights">The optional example weights.</param>
+        /// <param name="options">Algorithm advanced settings.</param>
+        /// <param name="onFit">A delegate that is called every time the
+        /// <see cref="Estimator{TInShape, TOutShape, TTransformer}.Fit(DataView{TInShape})"/> method is called on the
+        /// <see cref="Estimator{TInShape, TOutShape, TTransformer}"/> instance created out of this. This delegate will receive
+        /// the linear model that was trained.  Note that this action cannot change the result in any way; it is only a way for the caller to
+        /// be informed about what was learnt.</param>
+        /// <returns>The predicted output.</returns>
+        public static (Vector<float> score, Key<uint> predictedLabel) KMeans(this ClusteringContext.ClusteringTrainers ctx,
+           Vector<float> features, Scalar<float> weights,
```

Comment:
>Vector<float> features, Scalar<float> weights, [](start = 11, length = 46)

as i asked above, skip those?

Reply:
does [this ](https://github.com/dotnet/machinelearning/pull/2047#discussion_r246170641) comment from Tom help ?

Reply:
It is not directly answering my question, but i found my answer on point 4, here: [https://github.com/dotnet/machinelearning/issues/1798](https://github.com/dotnet/machinelearning/issues/1798) i think. 



---
In reply to: [248348858](https://github.com/dotnet/machinelearning/pull/2158#discussion_r248348858) [](ancestors = 248348858)

Reply:
Looking at this more, I think we can leave the static api as is; to avoid expanding the work. If we change the method signature, we'd need to also write new reconcilers for each task. 

But i think for the dynamic api, maybe we can skip the label, feature, weights and all other options from the second constructor that takes in the Options, to avoid having to make a decision about what overrides what, as stated in the original issue; link above. 

---
In reply to: [248359742](https://github.com/dotnet/machinelearning/pull/2158#discussion_r248359742) [](ancestors = 248359742,248348858)

Reply:
Agree. The API changes in this PR look good to me .  They also follow the the API  signatures  for SDCA (under review)  , FastTree  (checked-in)  . So we have consistency  

---
In reply to: [248366810](https://github.com/dotnet/machinelearning/pull/2158#discussion_r248366810) [](ancestors = 248366810,248359742,248348858)